### DISCUSSION TITLE
[docs] More clearly state the separation between vcpkg triplets and your project's settings

### DISCUSSION THREAD

**Discussion by tonygould** [microsoft/vcpkg#20742] [Q&A]
**Edit by @ras0219-msft:** I've modified the title of this issue to better reflect how it has changed from the original bug report into a docs improvement suggestion. The original report is as follows:

**Describe the bug**
Adding `set(VCPKG_TARGET_TRIPLET "x64-windows-static")` to CMakeLists.txt affects dependencies but not the current project.

**Environment**
- windows
- visual studio 2019

**To Reproduce**
Based on https://github.com/microsoft/vcpkg/blob/master/docs/examples/versioning.getting-started.md

CMakeLists.txt

    cmake_minimum_required(VERSION 3.18)

    set(VCPKG_TARGET_TRIPLET "x64-windows-static")
    set(VCPKG_HOST_TRIPLET "x64-windows-static")

    project(versionstest CXX)

    add_executable(main main.cpp)

    find_package(ZLIB REQUIRED)
    find_package(fmt CONFIG REQUIRED)
    target_link_libraries(main PRIVATE ZLIB::ZLIB fmt::fmt)

main.cpp

    #include <fmt/core.h>
    #include <zlib.h>

    int main() {
      fmt::print("fmt version is {}\n"
                "zlib version is {}\n",
                FMT_VERSION, ZLIB_VERSION);
      return 0;
    }

vcpkg.json

    {
        "name": "versions-test",
        "version": "1.0.0",
        "dependencies": [
            {
                "name": "fmt",
                "version>=": "7.1.3#4"
            },
            "zlib"
        ],
        "builtin-baseline": "b2544fd780b2e9a05d903314cbbd37268ba43d81"
    }

Dependent libraries build with /MT, but my library doesn't.

**Expected behavior**
Expected /MT to be applied everywhere (https://docs.microsoft.com/en-us/cpp/build/reference/md-mt-ld-use-run-time-library)

**Failure logs** 

config

	[main] Configuring folder: versiontest 
	[proc] Executing command: "c:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.EXE" --no-warn-unused-cli -DCMAKE_TOOLCHAIN_FILE:STRING=../vcpkg/scripts/buildsystems/vcpkg.cmake -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -Hc:/Users/anton/dev/versiontest -Bc:/Users/anton/dev/versiontest/build -G "Visual Studio 16 2019" -T host=x86 -A x64
	[cmake] Not searching for unused variables given on the command line.
	[cmake] -- Running vcpkg install
	[cmake] Detecting compiler hash for triplet x64-windows...
	[cmake] Detecting compiler hash for triplet x64-windows-static...
	[cmake] The following packages will be built and installed:
	[cmake]     fmt[core]:x64-windows-static -> 7.1.3#4 -- C:\Users\anton\dev\vcpkg\buildtrees\versioning\versions\fmt\230e140a15afbb9089537e153d8b83f5b994adbe
	[cmake]     vcpkg-cmake[core]:x64-windows -> 2021-02-28#2 -- C:\Users\anton\dev\vcpkg\buildtrees\versioning\versions\vcpkg-cmake\02fbd92ab46d1512b22280f1264e4e8f8d076ef4
	[cmake]     vcpkg-cmake-config[core]:x64-windows -> 2021-05-22 -- C:\Users\anton\dev\vcpkg\buildtrees\versioning\versions\vcpkg-cmake-config\2d4f997a32b8e8bfe98d12beb2bfe6be713c7086
	[cmake]     zlib[core]:x64-windows-static -> 1.2.11#10 -- C:\Users\anton\dev\vcpkg\buildtrees\versioning\versions\zlib\53a4615c8bb9b98a3864b834a6bbe51cc6c849ef
	[cmake] Using cached binary package: C:\Users\anton\AppData\Local\vcpkg\archives\b5\b50e4e5842d510da2661976301b0cf63178cf96d.zip
	[cmake] Using cached binary package: C:\Users\anton\AppData\Local\vcpkg\archives\f0\f0417292e76e7ca234c0a4641f8fc4eefd74540d.zip
	[cmake] Using cached binary package: C:\Users\anton\AppData\Local\vcpkg\archives\bf\bff24a6562b2bdfd785901a78d2eacbd0f841269.zip
	[cmake] Using cached binary package: C:\Users\anton\AppData\Local\vcpkg\archives\f5\f5efa3a2d5b9bc0722e6b93b84821874e8be39de.zip
	[cmake] Starting package 1/4: vcpkg-cmake-config:x64-windows
	[cmake] Building package vcpkg-cmake-config[core]:x64-windows...
	[cmake] Building package vcpkg-cmake-config[core]:x64-windows... done
	[cmake] Installing package vcpkg-cmake-config[core]:x64-windows...
	[cmake] Installing package vcpkg-cmake-config[core]:x64-windows... done
	[cmake] Elapsed time for package vcpkg-cmake-config:x64-windows: 26.09 ms
	[cmake] Starting package 2/4: vcpkg-cmake:x64-windows
	[cmake] Building package vcpkg-cmake[core]:x64-windows...
	[cmake] Building package vcpkg-cmake[core]:x64-windows... done
	[cmake] Installing package vcpkg-cmake[core]:x64-windows...
	[cmake] Installing package vcpkg-cmake[core]:x64-windows... done
	[cmake] Elapsed time for package vcpkg-cmake:x64-windows: 32.63 ms
	[cmake] Starting package 3/4: fmt:x64-windows-static
	[cmake] Building package fmt[core]:x64-windows-static...
	[cmake] Building package fmt[core]:x64-windows-static... done
	[cmake] Installing package fmt[core]:x64-windows-static...
	[cmake] Installing package fmt[core]:x64-windows-static... done
	[cmake] Elapsed time for package fmt:x64-windows-static: 130.5 ms
	[cmake] Starting package 4/4: zlib:x64-windows-static
	[cmake] Building package zlib[core]:x64-windows-static...
	[cmake] Building package zlib[core]:x64-windows-static... done
	[cmake] Installing package zlib[core]:x64-windows-static...
	[cmake] Installing package zlib[core]:x64-windows-static... done
	[cmake] Elapsed time for package zlib:x64-windows-static: 49.39 ms
	[cmake] 
	[cmake] Total elapsed time: 459.1 ms
	[cmake] 
	[cmake] -- Running vcpkg install - done
	[cmake] -- The CXX compiler identification is MSVC 19.29.30037.0
	[cmake] -- Detecting CXX compiler ABI info
	[cmake] -- Detecting CXX compiler ABI info - done
	[cmake] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30037/bin/Hostx86/x64/cl.exe - skipped
	[cmake] -- Detecting CXX compile features
	[cmake] -- Detecting CXX compile features - done
	[cmake] -- Found ZLIB: optimized;C:/Users/anton/dev/versiontest/build/vcpkg_installed/x64-windows-static/lib/zlib.lib;debug;C:/Users/anton/dev/versiontest/build/vcpkg_installed/x64-windows-static/debug/lib/zlibd.lib (found version "1.2.11") 
	[cmake] -- Configuring done
	[cmake] -- Generating done
	[cmake] -- Build files have been written to: C:/Users/anton/dev/versiontest/build

build

	[main] Building folder: versiontest 
	[build] Starting build
	[proc] Executing command: "c:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.EXE" --build c:/Users/anton/dev/versiontest/build --config Debug --target ALL_BUILD -j 10 --
	[build] Microsoft (R) Build Engine version 16.10.1+2fd48ab73 for .NET Framework
	[build] Copyright (C) Microsoft Corporation. All rights reserved.
	[build] 
	[build] fmtd.lib(format.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MTd_StaticDebug' doesn't match value 'MDd_DynamicDebug' in main.obj [C:\Users\anton\dev\versiontest\build\main.vcxproj]
	[build] libcpmtd.lib(locale0.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MTd_StaticDebug' doesn't match value 'MDd_DynamicDebug' in main.obj [C:\Users\anton\dev\versiontest\build\main.vcxproj]
	[build] libcpmtd.lib(locale0.obj) : error LNK2005: "void __cdecl std::_Facet_Register(class std::_Facet_base *)" (?_Facet_Register@std@@YAXPEAV_Facet_base@1@@Z) already defined in msvcprtd.lib(locale0_implib.obj) [C:\Users\anton\dev\versiontest\build\main.vcxproj]

(and lots more similar errors)

**Additional context**

I saw this problem first with a larger CMakePresets.json project, but made a repro in the similar case above. With CMakePresets.json, I'm using Ninja and the compiler line is

    [build] [11/125] "C:\PROGRA~2\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30037\bin\Hostx86\x64\cl.exe"  /nologo /TP   /DWIN32 /D_WINDOWS /EHsc /bigobj /Zi /Ob0 /Od /RTC1 -MDd -std:c++17 /showIncludes /Foxxxx.obj /Fdxxxx.cpp

when I expected -MT not -MD. Consistent with this, later in the build process, linking fails because there is a mismatch between my object files and those in the dependent libraries (which truly have been built with static per a CONTROL file I looked at -- "Architecture: x64-windows-static"). The preset here is

	"configurePresets": [
		{
			"name": "base",
			"description": "Sets generator, build and install directory, vcpkg",
			"hidden": true,
			"generator": "Ninja",
			"binaryDir": "${sourceDir}/out/build/${presetName}",
			"cacheVariables": {
				"CMAKE_INSTALL_PREFIX": "${sourceDir}/out/install/${presetName}",
				"CMAKE_TOOLCHAIN_FILE": {
					"type": "FILEPATH",
					"value": "../vcpkg/scripts/buildsystems/vcpkg.cmake"
				}
			},
			"environment": {
				"VCPKG_FEATURE_FLAGS": "manifests,versions"
			}
		},
		{
			"name": "base-windows",
			"inherits": [
				"base"
			],
			"hidden": true,
			"vendor": {
				"microsoft.com/VisualStudioSettings/CMake/1.0": {
					"hostOS": [
						"Windows"
					]
				}
			}
		},
		{
			"name": "msvc-debug",
			"inherits": [
				"base-windows"
			],
			"description": "Visual Studio debug",
			"cacheVariables": {
			   "CMAKE_BUILD_TYPE": "Debug",
				"CMAKE_C_COMPILER": "cl",
				"CMAKE_CXX_COMPILER": "cl",
				"VCPKG_TARGET_TRIPLET": "x64-windows-static",
				"VCPKG_HOST_TRIPLET": "x64-windows-static",
			}
		},

**Comment by Osyotr** [comment]
> Adding set(VCPKG_TARGET_TRIPLET "x64-windows-static") to CMakeLists.txt affects dependencies but not the current project.

Why would it? This triplet is created to reflect settings in your project and not to change them.
A warning would be cool, though.

**Comment by tonygould** [comment]
@Osyotr Thank you for your reply. I've clearly misunderstood how VCPKG_TARGET_TRIPLET works. Just to explain where I'm coming from, I first noticed this when I was building a large CMakePresets.json file, with different configs for msvc debug, msvc release, clang debug, gcc... etc (I've never used vcpkg in the old way of manually installing packages). My mindset was that each configuration in CMakePresets.json specifies a platform, a compiler, a triplet, and so on. I only tried changing a triplet in CMakeLists.txt to show what I thought was I bug. I can see now that of course VCPKG is about bring in packages so a setting beginning VCPKG_ is only going to affect dependencies and not the current project (as you say a warning would be nice, especially to cover a case where a genuine problem might not appear immediately -- switching to static gives some immediate warnings was just something I did as a test case).

OTOH, it would be great if there _were_ some way to do what I wanted. It does feel natural that in CMakePresets.json you could specify a triplet that would affect both the current and dependent packages. Some way of automatically getting the compiler linker flags etc to apply from a triplet in the current project would be helpful -- do you know of a way of doing this? In triplets/x64-windows-static.cmake, I see `set(VCPKG_CRT_LINKAGE static)`, but I can't see how to get from there to `/MTd` or `/MT`. Maybe I need to explore making my own package next.

**Comment by Osyotr** [comment]
@tonygould I'm probably wrong. After #9452 vcpkg sets `CMAKE_MSVC_RUNTIME_LIBRARY` to MT[d] for static triplets.
Perhaps you are using outdated CMake (<3.15)?

**Comment by tonygould** [comment]
@Osyotr Thanks for digging that out; I'm on 3.20 with `cmake_minimum_required(VERSION 3.18)` in the toy example above. I'm a little confused now as to what the expected behaviour is-- i.e. whether the current project _should_ follow VCPKG_TARGET_TRIPLET or not.

**Comment by Osyotr** [comment]
@tonygould did you also set CMP0091 to NEW?
> This policy was introduced in CMake version 3.15. Use the cmake_policy() command to set it to OLD or NEW explicitly. Unlike many policies, CMake version 3.20.4 does not warn when this policy is not set and simply uses OLD behavior.

**Comment by tonygould** [comment]
@Osyotr Thanks, I found https://cmake.org/cmake/help/git-stage/policy/CMP0091.html and also https://cmake.org/cmake/help/git-stage/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html ([stackoverflow](https://stackoverflow.com/a/56490614)). 

If in my CMakeLists.txt, I add the following, then the current project builds static

    cmake_policy(SET CMP0091 NEW)
    set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")

and everything compiles and links.

I guess what I expected to happen is for code like this to be in place

    cmake_policy(SET CMP0091 NEW)
    if ("${VCPKG_TARGET_TRIPLET}" STREQUAL "x64-windows-static")
        set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
    endif()

so that if I specify triplet `x64-windows-static` in CMakePresets.json, then the settings are applied to the current project (by default). Static is just a toy example -- in general I'd want triplet settings applied to dependencies and current project. But I can see why that isn't the case, at least historically, per your original reply, and as I'm not a package management expert by any stretch, there are likely some implications of doing that in the future that I'm missing. One thing I like about vcpkg compared to other package managers is that it doesn't try to do too much, so I can see the argument for not adding too many bells and whistles.

**Comment by JackBoosY** [comment]
reproducing...

**Comment by JackBoosY** [comment]
Can you please use triplet `x64-windows-static-md` instead?

**Comment by tonygould** [comment]
hi @JackBoosY I'm trying to test switching to static linking to the CRT -- that's what `x64-windows-static` does; `x64-windows-static-md` is different. Please take a look at the link to Microsoft help under "expected behaviour" in my original post, which explains the difference between /MT and /MD. They key issue is whether VCPKG_TARGET_TRIPLET should affect the current project; are you or one of your colleagues able to comment on this?

**Comment by JackBoosY** [comment]
@tonygould `VCPKG_TARGET_TRIPLET` defined the following variables:
- VCPKG_LIBRARY_LINKAGE
- VCPKG_CRT_LINKAGE
- VCPKG_TARGET_ARCHITECTURE
- VCPKG_CMAKE_SYSTEM_NAME
- VCPKG_CMAKE_SYSTEM_VERSION

It DOES effect the compiler option `/MT` or `/MD`.

**Comment by tonygould** [comment]
Hi @JackBister Yes, we know it affects the dependent projects (fmt, zlib). The question is whether it also affects the current project that I'm building, i.e. should it affect main.cpp compile and link in my example above? (Which is easy to replicate.)

**Comment by JackBoosY** [comment]
@tonygould I think it shouldn't affects the custom project, I will check that later.

**Comment by Osyotr** [comment]
@JackBoosY please see #9452

**Comment by strega-nil-ms** [comment]
This is an inversion of control that _we do not do_. It is not vcpkg's job to set up your CMake.

We expect that you will use an option, and set the triplet based on that option; for example:

```cmake
option(BUILD_SHARED_LIBS "Build shared libs instead of static libs" OFF)
option(SHARED_STANDARD_LIBRARY "Use /MD instead of /MT" ON)

if(NOT DEFINED VCPKG_TARGET_TRIPLET)
  set(default_vcpkg_arch ...)
  if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
    set(default_vcpkg_platform "windows")
    if(BUILD_SHARED_LIBS AND SHARED_STANDARD_LIBRARY)
      set(default_vcpkg_triplet_options "")
    elseif(SHARED_STANDARD_LIBRARY)
      set(default_vcpkg_triplet_options "static-md")
    elseif(NOT BUILD_SHARED_LIBS)
      set(default_vcpkg_triplet_options "static")
    else()
      message(FATAL_ERROR "Unsupported combination: static libs + /MD")
    endif()
  elseif(CMAKE_SYSTEM_NAME STREQUAL "Linux")
    # ...
  elseif(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
    # ...
  else()
    message(FATAL_ERROR "Unknown platform - set VCPKG_TARGET_TRIPLET: ${CMAKE_SYSTEM_NAME}")
  endif()
  set(VCPKG_TARGET_TRIPLET "${default_vcpkg_arch}-${default_vcpkg_platform}${default_vcpkg_triplet_options}")
endif()
```

**Comment by tonygould** [comment]
@strega-nil-ms That makes a lot of sense thanks. But it's not obvious to me starting to use vcpkg for the first time. Is there some documentation somewhere giving examples of use, or "how to" guidelines, like the one you gave that people would typically use in an enterprise setting? Especially w.r.t. the new manifests, versioning, binary caching etc?
